### ISSUE TITLE
Better #[do_not_trace] handling

### ISSUE THREAD

**Issue by bjorn3** [softdevteam/yk#244]
It shouldn't require the first block and return terminator of `#[do_not_trace]` function to be traced. This is important for software tracing. In case of hardware tracing it happens by accident as disabling hardware tracing is relatively costly, but in case of software tracing it did be nice to omit calls into the software tracer for `#[do_not_trace]` functions.

**Comment by ltratt** [comment]
Is this a correctness or a performance issue? I *think* it's just the latter?

**Comment by bjorn3** [comment]
There may be cases where it is really necessary to avoid the software tracing calls, but otherwise it is mostly a performance issue.

**Comment by ltratt** [comment]
Do we have a concrete example of such cases?

**Comment by bjorn3** [comment]
The only example I can think of is that inside interrupt handlers TLS accesses are UB. The software tracing callback uses TLS to determine if tracing is enabled. There may be more examples, but this is the only I can think of right now.

**Comment by ltratt** [comment]
> The only example I can think of is that inside interrupt handlers TLS accesses are UB

Right, I agree that we can't do software tracing inside interrupts / signals. I don't think we would do so, however, because the signal handler would point to the non-software-tracing version and thus would execute normal code. Warning: I might be wrong.

**Comment by bjorn3** [comment]
> I don't think we would do so, however, because the signal handler would point to the non-software-tracing version and thus would execute normal code. Warning: I might be wrong.

Every function is either traced or not. There is only ever one version being created. Without `#[do_not_trace]` all function in a crate are software traced when enabling software tracing support.

I just came up with another case where not tracing is important. When traced code calls a foreign function like from a C library and then this untraced code calls into a traced function, this part of the trace has to be trimmed.

* interp_step@main_exe
    * function_with_callback@some_c_lib
        * callback@main_exe

The TIR trace here would need to only contain the statements of `interp_step@main_exe` and a call to `function_with_callback@some_c_lib`. It must not contain `callback@main_exe`.

**Comment by vext01** [comment]
> When traced code calls a foreign function like from a C library and then this untraced code calls into a traced function.

And this also can occur in just Rust code too, where a `#[do_not_trace]` function then calls a traced function.

**Comment by ltratt** [comment]
Offline @vext01 has gently reminded me that my dreams do not always match reality :) In the long term, we want to have two versions of the interpreter in the binary:

* A fully optimised version in which a program starts execution.
* An unoptimised (or, perhaps, semi-optimised) version which we use for tracing.

We need the unoptimised version so that we can reliably relate a HW/SW trace back to the MIR: the optimised version will have moved things around too much for us to be able to do that.

A nice by-product of the above split is that we can make sure that things like interrupts always use the optimised version, which shouldn't do trace local stores etc.

However, at the moment we only have one version of the interpreter in the binary, so indeed we will write to memory at points that we shouldn't.

We *could* try and hack around this in the current set-up, which might be a bit tricky, or we could say "we'll wait until we achieve the above split and then the problem will go away anyway".

**Comment by bjorn3** [comment]
> or we could say "we'll wait until we achieve the above split and then the problem will go away anyway".

Makes sense. I guess I can just pretend that the `#[do_not_trace]` doesn't exist in cg_clif.

**Comment by ptersilie** [comment]
> And this also can occur in just Rust code too, where a `#[do_not_trace]` function then calls a traced function.

Does this matter? Since we just ignore anything between entering and leaving a `do_not_trace` function, and just turn it into a call, we don't care if that function calls other functions, traced or not traced. Or is this a software tracer specific problem?

**Comment by vext01** [comment]
I'm not sure actually :P

**Comment by bjorn3** [comment]
The problem here is that when a non-traced function calls a traced function, the traced function ends up in the final trace, but it must not be. The current ignore code works fine for hardware tracing in case of `#[do_not_trace]` as non-traced functions do still end up in the trace and as such it is possible to find the begin and end of the non-traced function, but in case of software tracing the non-traced function is completely missing, so we also don't know where it ends. This is also a problem for functions where SIR is completely missing (eg C function) calling traced functions, as the SIR return terminator is used to determine the end of non-traced functions.

**Comment by ptersilie** [comment]
That makes sense. I'm actually now wondering if this could also happen in hardware tracing. @vext01 Does the hardware tracer, when it sees a system call, completely stop tracing until it returns from that call (like what we do in TIR), or does it still trace stuff if that system call calls another function that can be traced?

**Comment by bjorn3** [comment]
On UNIX system calls never call other functions. That is only possible on Windows. (That is also why SEH unwinding is part of the kernel on Windows. Otherwise you can't unwind through the kernel.)

**Comment by ltratt** [comment]
I think there are, at least, two separate problems here:

1. How does software tracing know when things start/end?
2. The software tracer will do writes to memory in interrupt handlers, which is a correctness bug.

(2) I think, we'll just have to live with for now (see my suggestion in https://github.com/softdevteam/yk/issues/244#issuecomment-781216066 for the long term fix).

(1) doesn't look to me like a problem *except* if you have callbacks, which will certainly complicate the situation. We might be able to solve that problem by recording stack addresses at function entry/entry points, so that we can tell if we're in a callback or not.

**Comment by vext01** [comment]
> @vext01 Does the hardware tracer, when it sees a system call, completely stop tracing until it returns from that call

The `hwtracer` library flips a flag in the `perf_event_open(2)` options which causes PT to not report any blocks while in kernel space.

**Comment by bjorn3** [comment]
> How does software tracing know when things start/end?

It calls `__yk_swt_rec_loc` at the start of every block. Just before calling a `#[do_not_trace]` function it calls `__yk_swt_rec_func_addr`. This will result in a `SirLoc` for the first block of said function with it's address. The `#[do_not_trace]` however at a minimum also expects a `SirLoc` for the block containing the return terminator of said function. Maybe making `#[do_not_trace]` functions actually trace the entry block and return block would be enough to fix this?

**Comment by ltratt** [comment]
> Maybe making #[do_not_trace] functions actually trace the entry block and return block would be enough to fix this?

Yes, I think you might be right. I thought I had a counter example, but actually this approach solves that. Even if I pass a pointer to a function which is `do_not_trace`, provided we record its entry/exit blocks, we can do The Right Thing.

**Comment by bjorn3** [comment]
I will try it once #245 is merged, so I only have to rebase once.
